Atomic.md

```diff
@@ -0,0 +1,253 @@
+         the providerID, OnGetFileStream will fail so hydrating this
+         file will fail. Retrying will not help.
+
+-   Memcpy'ing the contentID xattr fails (or is never reached because of
+     a crash)
+
+    -   We have an incomplete placeholder file in the virtualization root
+        that cannot be recovered. There are no flags set on this file.
+
+
+    -   The MirrorProvider doesn't use contentID so no issues will occur
+
+    -   VFSForGit will be unable to get the SHA of the object that it
+         needs to download and hydrating this file will fail. Retrying
+         will not help.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   Placeholder does not have flags for virtualization root and
+        empty.
+
+    -   This file will never be expanded (but it will be the right size due to ftruncate) as the kext will exit when we
+         do not see the InVirtualizationRoot bit.
+
+-   We crash between setting the bits for 'in virtualization root' and 'is empty'
+
+    -   This file will never be expanded as the kext will exit
+            when we do not see the IsEmpty bit
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   The default mode should be 755 (unless a user has played with
+         umask, but this hole is out of scope for this design) so
+         hydration of the file should succeed. However, if the file had
+         a mode other than 755, tools that depend on it to be another
+         mode will fail. Recovery would require manual intervention by
+         chmodding this file.
+
+I believe that all of these problems can be addressed by moving to a
+model where we create placeholders in a temporary location, mark them up
+and then rename them into place.
+
+Proposed Solution:
+------------------
+
+This solution creates placeholders in a .staging directory within the
+provider's directory (.gvfs/.mirror), marks them up and then uses rename()
+to move them into the virtualization root.
+
+A sample of this change is implemented here (user mode component only):
+https://github.com/nickgra/VFSForGit/tree/atomic
+
+Currently, the provider registers its virtualization root with the kext
+at clone time. We have code today that will log on files detected
+outside of the virtualization root, which could be helpful for
+diagnosing customer issues.
+
+4974: 218.308) Note: No virtualization root found for file with set
+flag. (vnode path: \'/Users/nickgra/TestRoot/.staging/mov\')
+
+In the kext, we have widely assumed that files with `IsInVirtualizationRoot` will live inside our virtualization root. We will  fail to respond correctly to these kind of requests unless changes are made to handle the temporary directory scenario. Functionally, we want changes in registered temporary directories to be ignored. Changes will be needed on the provider side to register the
+temporary path alongside the kext changes to actually ignore activity in
+registered temp directories in the kext.
```

Comment:
Can we also use our kext to prevent the user from doing anything to our `.staging` folder? Ideally, limit _all_ IO to just the `gvfsmountuser` identity that we think we'll use to avoid the auth cache poison issues.